### DISCUSSION TITLE
A PDP-8/I in Antares

### DISCUSSION THREAD

**Discussion by fpw** [flandreas/antares#738] [Show and tell]
Hi all,

here's my first big project in Antares: Transcribing the original schematics of a [DEC PDP-8/I](https://en.wikipedia.org/wiki/PDP-8) into Antares!

![console](https://raw.githubusercontent.com/fpw/antares-pdp-8i/master/screenshots/console.png)

The project is published here as an Antares workspace: https://github.com/fpw/antares-pdp-8i

The PDP-8 is a computer from 1965 and considered the first minicomputer - that is, it only required a single cabinet and not an entire room! It is somewhat strange from today's perspective:

- the CPU doesn't use a clock, everything runs with manually calibrated delay lines. The speed was given as "333k additions per second"
- the architecture uses 12 bit words. Before ASCII became a thing, there were several different ways to encode characters. DEC's character set used 6 bit (opposed to 7/8 in ASCII), so it made sense that their architectures had 12, 18 and 36 bit. The term "byte" wasn't used back then.
- for that reason, octal made more sense than hexadecial - a 12-bit word can be encoded with 4 octal digits, one digit containing 3 bits. That's why everything on the panel is in groups of 3. The program counter in the screenshot above shows 2670 octal.
- bit indices are counted from the left, that is, the most-significant bit is assigned index 0. Nowadays, the convention is to count from the right, i.e. with the least-significant bit being 0. When a manual refers to SW0, it's the left-most switch.
- the machine has no builtin software (no boot ROM, no BIOS, no EFI...) so when turning it on initially, there is not even a way to load a program or output something. But programs can be toggled in using the front panel switches - you can directly manipulate the machine's RAM with the switches! The left-hand side of the front panel actually shows a short program that was used to load external programs from [paper tape](https://en.wikipedia.org/wiki/Punched_tape). But the memory was actually magnetic core memory: It's non-volatile, so once that program was toggled into the machine, it stayed there until you accidentally overwrote it, even when turning off the machine.
- while things like subroutine stacks were already invented, the PDP-8 doesn't have a stack. In fact, it doesn't even have a subtraction instruction. Everything was done in software. This was done to keep the PDP-8 small and cheap - and not because these things weren't already invented. It's basically a computer reduced down to the absolute bare necessities to keep it usable.

PDP-8s were used for many things that are now done using embedded computers: Controlling parts of nuclear power plants or devices like gas chromatographs, but also for office things like typesetting. Programming was commonly done in assembly, but there are compilers for COBOL, Pascal, Fortran and several other languages as well.

Since the complete schematics of the machine are available (see on project's page), this is an authentic circuit of a real computer running in Antares.

I'm not sure what kind of persons are reading this - if there is more interest, I could post some more information on how to use and program the machine.

**Comment by flandreas** [comment]
Words are not enough to express how beautifully this machine is done, and how proud I am that it is even possible to create something wonderful like this with Antares. When I started the Antares project 9 years ago, I wouldn't have believed it if someone told me that one day, someone would build such an amazing, gorgeous PDP-8 implementation with Antares.

Although I obviously don't understand yet how the PDP-8 works, I just enjoy clicking on the buttons, diving into the subcircuits, and watch the lamps going on and off :)

The first thing I've changed after loading the project was to set the property "Control Visibility" of the "Console" subcircuit in "Top" to "Always", so that the beautiful switches are also displayed in edit mode. Maybe you weren't aware of this property, or maybe you set it intentionally to "During Simulation", which is also fine.

And I've noticed that Antares doesn't properly display an hour glass mouse pointer when opening large circuits by double-clicking in the explorer, or when ALT double-clicking on subcircuit symbols to open it in a secondary view. I'll open an issue for that.

And finally, let me thank for your contributions to Antares in this last release period! You patiently hunted down old bugs, fixed them, and implemented new features. Thank you very much, it's a pleasure to cooperate with you!

**Reply by fpw** [reply]
Thanks for your kind words! :-)

Sub-Circuit visibility: Thanks, you were right that I didn't know about that - implemented that now.

Opening large circuits: I think there could be an alternate solution. I added a comment to the ticket that the slower loading times only happened after I used the new image feature, before that there was no difference between large and small circuits in loading time.

Finally, thanks for letting me contribute to Antares! In my initial survey of the currently available digital circuit simulation tools, it seemed to be the best tool for my undertaking and in the end it really proved to be the best choice!

**Comment by logoliv** [comment]
Impressive work !
